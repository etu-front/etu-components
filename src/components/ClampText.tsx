import React, { FC } from 'react'
import styled from 'styled-components'
import { BaseProps } from './types'

const Container = styled.div<{ title?: string, max: number }>`
  display: -webkit-box;
  line-clamp: ${props => props.max};
  -webkit-line-clamp: ${props => props.max};
  -webkit-box-orient: vertical;
  word-break: break-all;
  white-space: pre-wrap;
  overflow: hidden;
`

interface IProps {
  max?: number
  /**标题，当组件无 children 时，title 就为原文字 */
  title?: string
  text?: string
}
const ClampText: FC<IProps & BaseProps> = props => {
  const { max, title, text, className = '', children, style, ...rest } = props
  const clamp = Math.max(1, max || 1)
  return (
    <Container max={clamp} title={title || text} className={className}
      style={style}
      {...rest}
    >
      {children || title || text}
    </Container>
  )
}

export default ClampText
