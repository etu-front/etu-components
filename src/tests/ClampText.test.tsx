import React from 'react'
import renderer, { ReactTestRendererJSON } from 'react-test-renderer'
import 'jest-styled-components'
import { configure, mount } from 'enzyme'
import Adapter from '@cfaester/enzyme-adapter-react-18'

configure({ adapter: new Adapter() })

import ClampText from '../components/ClampText'

test('ClampText Renders and Click', () => {
  const mMock = jest.fn().mockImplementationOnce(() => 'first').mockImplementationOnce(() => 'two')

  const w = mount(<ClampText max={2} onClick={mMock} />)
  expect(w.find('div')).not.toBeNull()
  w.simulate('click')
  expect(mMock.mock.calls[0][0]).toBeTruthy()
  expect(mMock.mock.calls[0][0].target.tagName).toBe('DIV')
  expect(mMock()).toBe('two')
})

test('ClampText Component', () => {
  expect(ClampText.displayName).toEqual('ClampText')
})

test('ClampText 基本', () => {
  const component = renderer.create(<ClampText>hello</ClampText>)
  const tree = component.toJSON() as ReactTestRendererJSON
  expect(tree).not.toBeNull()
  expect(tree).toMatchSnapshot()
  if (!tree) return
  const { props } = tree
  expect(props).toHaveProperty('max', 1)
  expect(props.style).toBeUndefined()
  expect(props.title).toBeUndefined()
  expect(tree.children).toEqual(['hello'])
})

test('ClampText 复杂', () => {
  const component = renderer.create(
    <ClampText max={2} text="title" className="cls" style={{ lineHeight: 2 }}>hello</ClampText>
  )
  const tree = component.toJSON() as ReactTestRendererJSON
  expect(tree).not.toBeNull()
  expect(tree).toMatchSnapshot()
  if (!tree) return
  const { props } = tree
  expect(props).toHaveProperty('max', 2)
  expect(tree).toHaveStyleRule('line-clamp', '2')
  expect(props).toHaveProperty('title', 'title')
  expect(props.className).toContain('cls')
  expect(props.style).toHaveProperty('lineHeight', 2)
  expect(tree.children).toEqual(['hello'])
})

test('ClampText 标题', () => {
  const component = renderer.create(<ClampText max={3} text="title and children text" />)
  const tree = component.toJSON() as ReactTestRendererJSON
  expect(tree).not.toBeNull()
  expect(tree).toMatchSnapshot()
  if (!tree) return
  const { props } = tree
  expect(tree).toHaveStyleRule('line-clamp', '3')
  expect(props).toHaveProperty('max', 3)
  expect(props).toHaveProperty('title', 'title and children text')
  expect(tree.children).toEqual(['title and children text'])
})


