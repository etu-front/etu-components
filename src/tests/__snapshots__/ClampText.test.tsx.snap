// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClampText 基本 1`] = `
.c0 {
  display: -webkit-box;
  line-clamp: 1;
  -webkit-line-clamp: 1;
  -webkit-box-orient: vertical;
  word-break: break-all;
  white-space: pre-wrap;
  overflow: hidden;
}

<div
  className="c0"
  max={1}
>
  hello
</div>
`;

exports[`ClampText 复杂 1`] = `
.c0 {
  display: -webkit-box;
  line-clamp: 2;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  word-break: break-all;
  white-space: pre-wrap;
  overflow: hidden;
}

<div
  className="c0 cls"
  max={2}
  style={
    Object {
      "lineHeight": 2,
    }
  }
  title="title"
>
  hello
</div>
`;

exports[`ClampText 标题 1`] = `
.c0 {
  display: -webkit-box;
  line-clamp: 3;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
  word-break: break-all;
  white-space: pre-wrap;
  overflow: hidden;
}

<div
  className="c0"
  max={3}
  title="title and children text"
>
  title and children text
</div>
`;
